### PR TITLE
Rework connection manager ui

### PR DESCRIPTION
**Pull request by AlexanderSher**
#2222 Connection management
#2207 Endless UriFormatException
#2211 Connection Manager has the wrong title type:bug

### CODE REVIEW

File: src/Package/Impl/Resources.resx [hunk 1]
```diff
@@ -465,6 +465,9 @@
   <data name="HistoryWindowCaption" xml:space="preserve">
     <value>R History</value>
   </data>
+  <data name="ConnectionManagerWindowCaption" xml:space="preserve">
```
Comment (MikhailArkhipov) [inline src/Package/Impl/Resources.resx]: I think it is supposed to be 'Workspaces'

File: src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs [hunk 2]
```diff
@@ -73,50 +82,100 @@ internal sealed class ConnectionManagerViewModel : BindableBase, IConnectionMana
...
+            CancelEdit(SelectedConnection);
+            NewConnection = new ConnectionViewModel();
+        }
+        
+        public void CancelEdit(IConnectionViewModel connection) {
+            _shell.AssertIsOnMainThread();
+            if (connection == NewConnection) {
+                NewConnection = null;
+            } else if (connection != null) {
+                connection.Reset();
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs]: Reset() should probably auto-clear Editing

File: src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs [hunk 3]
```diff
@@ -73,50 +82,100 @@ internal sealed class ConnectionManagerViewModel : BindableBase, IConnectionMana
...
+            string latestLocalPath;
+            Uri latestLocalPathUri;
+
+            if (connection.Path != null && Uri.TryCreate(connection.Path, UriKind.Absolute, out latestLocalPathUri) && latestLocalPathUri.IsFile && !latestLocalPathUri.IsUnc) {
+                latestLocalPath = latestLocalPathUri.LocalPath;
+            } else { 
+                latestLocalPath = Environment.SystemDirectory;
+
+                try {
+                    latestLocalPath = new RInstallation().GetCompatibleEnginePathFromRegistry();
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs]: if it came from ChooseRFolderUIEditor class, then that class can be deleted I think

File: src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs [hunk 4]
```diff
@@ -73,50 +82,100 @@ internal sealed class ConnectionManagerViewModel : BindableBase, IConnectionMana
...
+            connection.IsEditing = true;
         }
 
-        public void SaveSelected() {
+        public Task TestConnectionAsync(IConnectionViewModel connection) => Task.CompletedTask;
+
+        public void Save(IConnectionViewModel connectionViewModel) {
             _shell.AssertIsOnMainThread();
-            if (string.IsNullOrEmpty(SelectedConnection.Name)) {
+            if (string.IsNullOrEmpty(connectionViewModel.Name)) {
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs]: Technically we can just disable Save button when there is no name

File: src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs [hunk 5]
```diff
@@ -73,50 +82,100 @@ internal sealed class ConnectionManagerViewModel : BindableBase, IConnectionMana
...
                 _shell.ShowMessage(Resources.ConnectionManager_ShouldHaveName, MessageButtons.OK);
                 return;
             }
 
-            if (string.IsNullOrEmpty(SelectedConnection.Path)) {
+            if (string.IsNullOrEmpty(connectionViewModel.Path)) {
                 _shell.ShowMessage(Resources.ConnectionManager_ShouldHavePath, MessageButtons.OK);
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs]: Same as above

File: src/R/Components/Impl/Resources.resx [hunk 6]
```diff
@@ -316,19 +316,22 @@
     <value>Connect to '{0}'.</value>
   </data>
   <data name="ConnectionManager_Name" xml:space="preserve">
-    <value>Name:</value>
+    <value>Enter the name of the connection.</value>
   </data>
   <data name="ConnectionManager_Path" xml:space="preserve">
-    <value>Path:</value>
+    <value>Enter the path to the connection. Either an URL or a local path.</value>
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/Resources.resx]: I'd put actual value such as ```http://machine:port or path to local R```

File: src/R/Components/Impl/Resources.resx [hunk 7]
```diff
@@ -316,19 +316,22 @@
...
   </data>
   <data name="ConnectionManager_RCommandLineArguments" xml:space="preserve">
-    <value>R Command Line Arguments:</value>
+    <value>Enter command line arguments that will be used when R starts.</value>
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/Resources.resx]: Just ```R command line arguments```

File: src/R/Components/Impl/Resources.resx [hunk 8]
```diff
@@ -316,19 +316,22 @@
     <value>Connect to '{0}'.</value>
   </data>
   <data name="ConnectionManager_Name" xml:space="preserve">
-    <value>Name:</value>
+    <value>Enter the name of the connection.</value>
```
Comment (MikhailArkhipov) [inline src/R/Components/Impl/Resources.resx]: ```Connection name```

**Comment by MikhailArkhipov** [comment]
:+1:

File: src/Package/Impl/Resources.resx (hunk 1 above)
Comment (AlexanderSher) [inline src/Package/Impl/Resources.resx]: Done

---
In reply to: [77055282](https://github.com/Microsoft/RTVS/pull/2237#discussion_r77055282) [](ancestors = 77055282)

File: src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs (hunk 2 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs]: Agreed.

---
In reply to: [77055531](https://github.com/Microsoft/RTVS/pull/2237#discussion_r77055531) [](ancestors = 77055531)

File: src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs (hunk 3 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs]: Yes it is. Deleted.

---
In reply to: [77056071](https://github.com/Microsoft/RTVS/pull/2237#discussion_r77056071) [](ancestors = 77056071)

File: src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs (hunk 4 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ViewModel/ConnectionManagerViewModel.cs]: Yes, it is disabled already. 
Moved strings to the save button tooltip

---
In reply to: [77056353](https://github.com/Microsoft/RTVS/pull/2237#discussion_r77056353) [](ancestors = 77056353)

**Pull request merged**